File: src/main/java/com/nugurang/graphql/resolver/ArticleResolver.java
package com.nugurang.graphql.resolver;

import com.nugurang.dao.ArticleDao;
import com.nugurang.dao.ImageDao;
import com.nugurang.dao.ThreadDao;
import com.nugurang.dao.UserDao;
import com.nugurang.dao.VoteDao;
import com.nugurang.dao.VoteTypeDao;
import com.nugurang.dto.ArticleDto;
import com.nugurang.dto.ImageDto;
import com.nugurang.dto.TagDto;
import com.nugurang.dto.ThreadDto;
import com.nugurang.dto.UserDto;
import graphql.kickstart.tools.GraphQLResolver;
import java.util.List;
import java.util.Optional;
import java.util.stream.Collectors;
import org.springframework.stereotype.Service;

@Service
public class ArticleResolver implements GraphQLResolver<ArticleDto> {
    private final ArticleDao articleDao;
    private final ImageDao imageDao;
    private final ThreadDao threadDao;
    private final UserDao userDao;
    private final VoteDao voteDao;
    private final VoteTypeDao voteTypeDao;

    public ThreadDto thread(ArticleDto articleDto) {
        return articleDao.findById(articleDto.getId()).map(articleEntity -> articleEntity.getThread()).map(entity -> entity.toDto()).get();
    }

    public UserDto user(ArticleDto articleDto) {
        return articleDao.findById(articleDto.getId()).map(articleEntity -> articleEntity.getUser()).map(userEntity -> userEntity.toDto()).get();
    }

    public Optional<ArticleDto> parent(ArticleDto articleDto) {
        return articleDao.findById(articleDto.getId()).map(articleEntity -> articleEntity.getParent()).map(entity -> entity.toDto());
    }

    public List<ImageDto> images(ArticleDto articleDto) {
        return imageDao.findAllByArticleId(articleDto.getId()).stream().map(entity -> entity.toDto()).collect(Collectors.toList());
    }

    public List<TagDto> tags(ArticleDto articleDto) {
        return null;
    }

    public Long viewCount(ArticleDto articleDto) {
        return voteDao.countByArticleIdAndVoteTypeId(articleDto.getId(), voteTypeDao.findByName("VIEW").getId());
    }

    public Long upCount(ArticleDto articleDto) {
        return voteDao.countByArticleIdAndVoteTypeId(articleDto.getId(), voteTypeDao.findByName("UP").getId());
    }

    public Long downCount(ArticleDto articleDto) {
        return voteDao.countByArticleIdAndVoteTypeId(articleDto.getId(), voteTypeDao.findByName("DOWN").getId());
    }

    public Long starCount(ArticleDto articleDto) {
        return voteDao.countByArticleIdAndVoteTypeId(articleDto.getId(), voteTypeDao.findByName("STAR").getId());
    }

    //<editor-fold defaultstate="collapsed" desc="delombok">
    @SuppressWarnings("all")
    
    public ArticleResolver(final ArticleDao articleDao, final ImageDao imageDao, final ThreadDao threadDao, final UserDao userDao, final VoteDao voteDao, final VoteTypeDao voteTypeDao) {
        this.articleDao = articleDao;
        this.imageDao = imageDao;
        this.threadDao = threadDao;
        this.userDao = userDao;
        this.voteDao = voteDao;
        this.voteTypeDao = voteTypeDao;
    }
    //</editor-fold>
}

File: src/main/kotlin/com/nugurang/graphql/resolver/ArticleResolver.kt
package com.nugurang.graphql.resolver

import com.nugurang.dao.ArticleDao
import com.nugurang.dao.ImageDao
import com.nugurang.dao.VoteDao
import com.nugurang.dao.VoteTypeDao
import com.nugurang.dto.*
import com.nugurang.entity.ThreadEntity
import com.nugurang.entity.UserEntity
import com.nugurang.entity.VoteTypeEntity
import com.nugurang.exception.NotFoundException
import com.nugurang.mapper.ArticleMapper
import com.nugurang.mapper.ImageMapper
import graphql.kickstart.tools.GraphQLResolver
import org.springframework.data.repository.findByIdOrNull
import org.springframework.stereotype.Service

@Service
class ArticleResolver(
    private val articleDao: ArticleDao,
    private val imageDao: ImageDao,
    private val voteDao: VoteDao,
    private val voteTypeDao: VoteTypeDao,
    private val articleMapper: ArticleMapper,
    private val imageMapper: ImageMapper
) : GraphQLResolver<ArticleDto> {

    fun thread(articleDto: ArticleDto): ThreadDto {
        return articleDao.findByIdOrNull(articleDto.id)?.thread?.toDto()
        ?: throw NotFoundException(ThreadEntity::class.java)
    }

    fun user(articleDto: ArticleDto): UserDto {
        return articleDao.findByIdOrNull(articleDto.id)?.user?.toDto()
        ?: throw NotFoundException(UserEntity::class.java)
    }

    // TODO: should we return nullable or non-nullable with an exception?
    fun parent(articleDto: ArticleDto): ArticleDto? {
        return articleDao.findByIdOrNull(articleDto.id)?.parent?.let(articleMapper::toDto)
    }

    fun images(articleDto: ArticleDto): List<ImageDto> {
        return imageDao.findAllByArticleId(articleDto.id).map(imageMapper::toDto)
    }

    fun tags(articleDto: ArticleDto): List<TagDto> {
        return listOf()
    }

    fun viewCount(articleDto: ArticleDto): Long {
        return voteDao.countByArticleIdAndVoteTypeId(
            articleDto.id,
            voteTypeDao.findByName("VIEW")?.id
            ?: throw NotFoundException(VoteTypeEntity::class.java)
        )
    }

    fun upCount(articleDto: ArticleDto): Long {
        return voteDao.countByArticleIdAndVoteTypeId(
            articleDto.id,
            voteTypeDao.findByName("UP")?.id
            ?: throw NotFoundException(VoteTypeEntity::class.java)
        )
    }

    fun downCount(articleDto: ArticleDto): Long {
        return voteDao.countByArticleIdAndVoteTypeId(
            articleDto.id,
            voteTypeDao.findByName("DOWN")?.id
            ?: throw NotFoundException(VoteTypeEntity::class.java)
        )
    }

    fun starCount(articleDto: ArticleDto): Long {
        return voteDao.countByArticleIdAndVoteTypeId(
            articleDto.id,
            voteTypeDao.findByName("STAR")?.id
            ?: throw NotFoundException(VoteTypeEntity::class.java)
        )
    }
}